compiler/runtime/OMRRelocationRecord.cpp

```diff
@@ -0,0 +1,399 @@
+   }
+
+int32_t
+OMR::RelocationRecordGroup::applyRelocations(TR::RelocationRuntime *reloRuntime,
+                                           TR::RelocationTarget *reloTarget,
+                                           uint8_t *reloOrigin)
+   {
+   TR::RelocationRecordBinaryTemplate *recordPointer = firstRecord(reloTarget);
+   TR::RelocationRecordBinaryTemplate *endOfRecords = pastLastRecord(reloTarget);
+
+   while (recordPointer < endOfRecords)
+      {
+      TR::RelocationRecord storage;
+      // Create a specific type of relocation record based on the information
+      // in the binary record pointed to by `recordPointer`
+      TR::RelocationRecord *reloRecord = TR::RelocationRecord::create(&storage, reloRuntime, reloTarget, recordPointer);
+      int32_t rc = handleRelocation(reloRuntime, reloTarget, reloRecord, reloOrigin);
+      if (rc != 0)
+         return rc;
+
+      recordPointer =  reinterpret_cast<TR::RelocationRecordBinaryTemplate *>(reloRecord->nextBinaryRecord(reloTarget));
+      }
+
+   return 0;
+   }
+
+void OMR::RelocationRecord::initialize(TR::RelocationTarget *reloTarget,uint8_t* targetAddress, uint8_t* targetAddress2 )
+   {
+   TR_UNIMPLEMENTED();
+   }
+
+int32_t
+OMR::RelocationRecordGroup::handleRelocation(TR::RelocationRuntime *reloRuntime,
+                                           TR::RelocationTarget *reloTarget,
+                                           TR::RelocationRecord *reloRecord,
+                                           uint8_t *reloOrigin)
+   {
+   reloRecord->preparePrivateData(reloRuntime, reloTarget);
+   return reloRecord->applyRelocationAtAllOffsets(reloRuntime, reloTarget, reloOrigin);
+   }
+
+#define FLAGS_RELOCATION_WIDE_OFFSETS   0x80
+#define FLAGS_RELOCATION_EIP_OFFSET     0x40
+#define FLAGS_RELOCATION_TYPE_MASK      (TR::ExternalRelocationTargetKindMask)
+#define FLAGS_RELOCATION_FLAG_MASK      ((uint8_t) (FLAGS_RELOCATION_WIDE_OFFSETS | FLAGS_RELOCATION_EIP_OFFSET))
+
+OMR::RelocationRecordAction
+OMR::RelocationRecord::action(OMR::RelocationRuntime *reloRuntime)
+   {
+   return RelocationRecordAction::apply;
+   }
+
+
+TR::RelocationRecord *
+TR::RelocationRecord::create(TR::RelocationRecord *storage, TR::RelocationRuntime *reloRuntime, TR::RelocationTarget *reloTarget, TR::RelocationRecordBinaryTemplate *record)
+   {
+   OMR::RelocationRecord *reloRecord = NULL;
+   // based on the type of the relocation record, create an object of a particular variety of OMR::RelocationRecord object
+   uint8_t reloType = record->type(reloTarget);
+   switch (reloType)
+      {
+      default:
+         // TODO: error condition
+         TR_UNIMPLEMENTED();
+         exit(0);
+      }
+   
+   return static_cast<TR::RelocationRecord*>(reloRecord);
+   }
+
+void
+OMR::RelocationRecord::print(TR::RelocationRuntime *reloRuntime)
+   {
+   TR::RelocationTarget *reloTarget = reloRuntime->reloTarget();
```

Comment:
not sure why this line of code is here...

Reply:
Thank you for the review! The plan is to introduce infrastructural changes by putting the files here, that would not interfere with the product behavior, and do the incremental improvements after that. First focus being the functional verification.

Reply:
Thank you for the review! The plan is to introduce infrastructural changes by putting the files here, that would not interfere with the product behavior, and do the incremental improvements after that. First focus being the functional verification.